File: scripts/benchmark_gateway_flight.jl
using Statistics
using Tables
using WendaoArrow

Base.@kwdef struct GatewayFlightBenchConfig
    host::String = WendaoArrow.DEFAULT_HOST
    port::Int = WendaoArrow.DEFAULT_GATEWAY_FLIGHT_PORT
    query::String = "flight"
    limit::Int = 5
    samples::Int = 10
    route::Symbol = :both
end

function parse_bench_args(args::Vector{String})
    host = WendaoArrow.DEFAULT_HOST
    port = WendaoArrow.DEFAULT_GATEWAY_FLIGHT_PORT
    query = "flight"
    limit = 5
    samples = 10
    route = :both
    index = 1
    while index <= length(args)
        argument = args[index]
        if startswith(argument, "--host=")
            host = split(argument, "=", limit = 2)[2]
        elseif startswith(argument, "--port=")
            port = parse(Int, split(argument, "=", limit = 2)[2])
        elseif startswith(argument, "--query=")
            query = split(argument, "=", limit = 2)[2]
        elseif startswith(argument, "--limit=")
            limit = parse(Int, split(argument, "=", limit = 2)[2])
        elseif startswith(argument, "--samples=")
            samples = parse(Int, split(argument, "=", limit = 2)[2])
        elseif startswith(argument, "--route=")
            route = Symbol(split(argument, "=", limit = 2)[2])
        elseif argument in
               ("--host", "--port", "--query", "--limit", "--samples", "--route")
            index += 1
            index > length(args) && throw(ArgumentError("missing value for $(argument)"))
            value = args[index]
            if argument == "--host"
                host = value
            elseif argument == "--port"
                port = parse(Int, value)
            elseif argument == "--query"
                query = value
            elseif argument == "--limit"
                limit = parse(Int, value)
            elseif argument == "--samples"
                samples = parse(Int, value)
            else
                route = Symbol(value)
            end
        else
            throw(ArgumentError("unsupported benchmark argument: $(argument)"))
        end
        index += 1
    end
    samples > 0 || throw(ArgumentError("benchmark samples must be greater than zero"))
    limit > 0 || throw(ArgumentError("benchmark limit must be greater than zero"))
    route in (:repo, :knowledge, :both) || throw(
        ArgumentError(
            "benchmark route must be one of :repo, :knowledge, or :both; got $(route)",
        ),
    )
    return GatewayFlightBenchConfig(;
        host = host,
        port = port,
        query = query,
        limit = limit,
        samples = samples,
        route = route,
    )
end

function bench_case(label::AbstractString, fetch::Function; samples::Int)
    warm_table = fetch()
    row_count = length(Tables.rows(warm_table))
    column_count = length(Tables.schema(warm_table).names)
    times_ms = Float64[]
    alloc_bytes = Int[]
    for _ = 1:samples
        GC.gc()
        result = @timed fetch()
        push!(times_ms, result.time * 1000)
        push!(alloc_bytes, result.bytes)
    end
    return (
        case = label,
        median_ms = median(times_ms),
        minimum_ms = minimum(times_ms),
        maximum_ms = maximum(times_ms),
        alloc_bytes = median(alloc_bytes),
        rows = row_count,
        columns = column_count,
    )
end

function print_results(results)
    println("case\tmedian_ms\tminimum_ms\tmaximum_ms\talloc_bytes\trows\tcolumns")
    for result in results
        println(
            join(
                (
                    result.case,
                    string(round(result.median_ms; digits = 3)),
                    string(round(result.minimum_ms; digits = 3)),
                    string(round(result.maximum_ms; digits = 3)),
                    string(result.alloc_bytes),
                    string(result.rows),
                    string(result.columns),
                ),
                '\t',
            ),
        )
    end
end

function main(args::Vector{String})
    config = parse_bench_args(args)
    client = WendaoArrow.gateway_flight_client(;
        host = config.host,
        port = config.port,
        deadline = 30,
    )
    results = NamedTuple[]
    if config.route in (:repo, :both)
        push!(
            results,
            bench_case(
                "repo_search",
                () -> WendaoArrow.gateway_repo_search(
                    client,
                    config.query;
                    limit = config.limit,
                );
                samples = config.samples,
            ),
        )
    end
    if config.route in (:knowledge, :both)
        push!(
            results,
            bench_case(
                "knowledge_search",
                () -> WendaoArrow.gateway_knowledge_search(
                    client,
                    config.query;
                    limit = config.limit,
                );
                samples = config.samples,
            ),
        )
    end
    print_results(results)
    return nothing
end

main(ARGS)

File: scripts/benchmark_gateway_flight.py
from __future__ import annotations

import argparse
import time

import pyarrow.flight as flight

DEFAULT_HOST = "127.0.0.1"
DEFAULT_GATEWAY_FLIGHT_PORT = 9517
DEFAULT_GATEWAY_SCHEMA_VERSION = "v2"
DEFAULT_GATEWAY_RESULT_LIMIT = 10
DEFAULT_GATEWAY_REPO_SEARCH_ROUTE = ("search", "repos", "main")
DEFAULT_GATEWAY_KNOWLEDGE_SEARCH_ROUTE = ("search", "knowledge")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Benchmark one running Rust gateway Flight surface over pyarrow.flight.",
    )
    parser.add_argument("--host", default=DEFAULT_HOST)
    parser.add_argument("--port", type=int, default=DEFAULT_GATEWAY_FLIGHT_PORT)
    parser.add_argument("--query", default="flight")
    parser.add_argument("--limit", type=int, default=5)
    parser.add_argument("--samples", type=int, default=10)
    parser.add_argument("--route", choices=("repo", "knowledge", "both"), default="both")
    return parser.parse_args()


def median(values: list[float]) -> float:
    if not values:
        return 0.0
    sorted_values = sorted(values)
    midpoint = len(sorted_values) // 2
    if len(sorted_values) % 2:
        return sorted_values[midpoint]
    return (sorted_values[midpoint - 1] + sorted_values[midpoint]) / 2.0


def base_headers() -> list[tuple[bytes, bytes]]:
    return [(b"x-wendao-schema-version", DEFAULT_GATEWAY_SCHEMA_VERSION.encode("utf-8"))]


def repo_headers(query: str, limit: int) -> list[tuple[bytes, bytes]]:
    return [
        *base_headers(),
        (b"x-wendao-repo-search-query", query.encode("utf-8")),
        (b"x-wendao-repo-search-limit", str(limit).encode("utf-8")),
    ]


def knowledge_headers(query: str, limit: int) -> list[tuple[bytes, bytes]]:
    return [
        *base_headers(),
        (b"x-wendao-search-query", query.encode("utf-8")),
        (b"x-wendao-search-limit", str(limit).encode("utf-8")),
    ]


def call_options(headers: list[tuple[bytes, bytes]]) -> flight.FlightCallOptions:
    return flight.FlightCallOptions(timeout=30.0, headers=headers)


def read_route_table(
    client: flight.FlightClient,
    route_segments: tuple[str, ...],
    headers: list[tuple[bytes, bytes]],
):
    descriptor = flight.FlightDescriptor.for_path(*route_segments)
    options = call_options(headers)
    info = client.get_flight_info(descriptor, options)
    endpoint = info.endpoints[0]
    reader = client.do_get(endpoint.ticket, options)
    return reader.read_all()


def bench_case(
    client: flight.FlightClient,
    *,
    label: str,
    route_segments: tuple[str, ...],
    headers: list[tuple[bytes, bytes]],
    samples: int,
) -> dict[str, object]:
    warm_table = read_route_table(client, route_segments, headers)
    times_ms: list[float] = []
    for _ in range(samples):
        started = time.perf_counter_ns()
        _ = read_route_table(client, route_segments, headers)
        times_ms.append((time.perf_counter_ns() - started) / 1_000_000.0)
    return {
        "case": label,
        "median_ms": median(times_ms),
        "minimum_ms": min(times_ms),
        "maximum_ms": max(times_ms),
        "rows": warm_table.num_rows,
        "columns": warm_table.num_columns,
    }


def print_results(results: list[dict[str, object]]) -> None:
    print("case\tmedian_ms\tminimum_ms\tmaximum_ms\trows\tcolumns")
    for result in results:
        print(
            "\t".join(
                (
                    str(result["case"]),
                    f"{float(result['median_ms']):.3f}",
                    f"{float(result['minimum_ms']):.3f}",
                    f"{float(result['maximum_ms']):.3f}",
                    str(result["rows"]),
                    str(result["columns"]),
                )
            )
        )


def main() -> int:
    args = parse_args()
    if not args.host.strip():
        raise ValueError("benchmark host must be non-empty")
    if args.port <= 0:
        raise ValueError("benchmark port must be greater than zero")
    if not args.query.strip():
        raise ValueError("benchmark query must be non-empty")
    if args.limit <= 0:
        raise ValueError("benchmark limit must be greater than zero")
    if args.samples <= 0:
        raise ValueError("benchmark samples must be greater than zero")

    client = flight.connect(f"grpc://{args.host}:{args.port}")
    results: list[dict[str, object]] = []
    if args.route in ("repo", "both"):
        results.append(
            bench_case(
                client,
                label="repo_search",
                route_segments=DEFAULT_GATEWAY_REPO_SEARCH_ROUTE,
                headers=repo_headers(args.query, args.limit),
                samples=args.samples,
            )
        )
    if args.route in ("knowledge", "both"):
        results.append(
            bench_case(
                client,
                label="knowledge_search",
                route_segments=DEFAULT_GATEWAY_KNOWLEDGE_SEARCH_ROUTE,
                headers=knowledge_headers(args.query, args.limit),
                samples=args.samples,
            )
        )
    close = getattr(client, "close", None)
    if callable(close):
        close()
    print_results(results)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
